src/client/dfuse/ops/readdir.c

```diff
@@ -235,8 +236,12 @@ dfuse_cb_readdir(fuse_req_t req, struct dfuse_obj_hdl *oh,
 	/* if starting from the beginning, reset the anchor attached to
 	 * the open handle.
 	 */
-	if (offset == 0)
+	if (offset == 0) {
+		if (oh->doh_kreaddir_started)
+			oh->doh_kreaddir_invalid = true;
```

Comment:
Is this ever set to `false`?

Reply:
It defaults to false as it's a bool, once you set it to true then it remains true.  What this means is practice is that the application did a seekdir() back to the beginning and if this happens then the kernel will not populate the cache from this directory handle so we shouldn't mark it as being a potential cache save.

Mostly this PR is modelling if the kernel will cache the data or not, and if dfuse thinks it might have done then apply timeouts based on that, there is no clear API for querying the contents of the cache or seeing when it was populated.  dfuse could try reading the cache data in both opendir and closedir to get a (potentially) more accurate picture however that would add significant overhead as you actually need to read the cache data itself rather than querying it's validity.